[Merged by Bors] - feat(Geometry/Diffeology): basics of diffeological spaces

peabrainiac opened the pull request:
Introduces diffeological spaces, smooth maps between them, the D-topology and the standard diffeology on finite-dimensional normed spaces.

---

This PR continues the work from #21969.

peabrainiac committed:
copy file from orbifolds repo

peabrainiac committed:
module docstring & some formatting changes

peabrainiac committed:
import in Mathlib.lean

peabrainiac committed:
add missing docstrings

peabrainiac committed:
add references & tags

peabrainiac committed:
fix indentation

peabrainiac committed:
minor adjustments

peabrainiac committed:
merge master

peabrainiac committed:
rename `ContDiffCompatible` & `DTopCompatible`

peabrainiac committed:
add `fun_prop` lemmas

peabrainiac committed:
update noshake.json

peabrainiac committed:
update comments

peabrainiac committed:
merge master

peabrainiac committed:
use `obtain` instead of `cases'`

peabrainiac committed:
merge master

peabrainiac committed:
fix

peabrainiac committed:
merge master

peabrainiac committed:
fix

peabrainiac commented:
@grunweg I've merged master and migrated the PR now; sorry for taking a bit longer to do so. I'm looking forward to your feedback once you find the time.

@lecopivo there appears to have been some breakage with `fun_prop` when merging master just now; it seems to fail to infer `DSmooth p` from `IsPlot p` where previously that worked without a problem (see the second change in the latest commit / the bottom of the file). Is that a bug / regression in `fun_prop` or am I doing something wrong?

lecopivo commented on Mathlib/Geometry/Diffeology/Basic.lean:
```diff
@@ -0,0 +1,460 @@
+@[fun_prop]
+protected theorem IsPlot.dsmooth {n : ℕ} {p : Eucl n → X} (hp : IsPlot p) : DSmooth p :=
+  fun _ _ ↦ isPlot_reparam hp
+
```
@JovanGerb There seems to be an issue with `RefinedDiscrTree` and it is having a hard time looking up this theorem.

This piece of meta code looks up transition theorems for `DSmooth p` and it won't give back `IsPlot.dsmooth` with `zeta:=false`. 
```
open Mathlib.Meta.FunProp Lean Meta Qq in
run_meta
  withLocalDeclDQ `f q(Eucl 3 → Eucl 3) fun f => do
   let thms := (transitionTheoremsExt.getState (← getEnv)).theorems
   let (candidates, thms) ← withConfig (fun cfg => { cfg with iota := false, zeta := true }) <|
     thms.getMatch q(DSmooth $f) false true
   let candidates := (← MonadExcept.ofExcept candidates).toArray
   for c in candidates do
     IO.println c.thmName
```

Could it be related to the new refactor of how `let/have/letFun` works?

lecopivo replied:
I will try to produce a mwe that does not depend on this PR. Unfortunately a trivial attempt didn't work and something more complicated is at work.

JovanGerb replied:
I have now verified that on #26531 the above failure is indeed fixed.

peabrainiac committed:
remove old noshake.json entry

peabrainiac committed:
merge master

peabrainiac committed:
restore previously broken `fun_prop` test case

peabrainiac commented:
@grunweg since #26531 is now merged, this PR should now be ready for review again - I've already merged master and restored the previously broken application of `fun_prop` to its original state. I've also now opened a third PR in this chain of PRs, building further towards constructions of diffeological spaces.

grunweg commented:
Thanks! I'll tentatively assign this PR to me --- but you should know that I'll be on holiday next week, travelling the week after and on holiday again during the last week of September: it is not certain whether I'll get to this before October.

peabrainiac committed:
merge master

peabrainiac committed:
remove `Eucl n` abbrev for `EuclideanSpace ℝ (Fin n)`

peabrainiac committed:
add `assert_not_exists` for manifold stuff

ocfnash reviewed (commented):
I'm sorry we've been neglecting this PR for so long. I'd like to get this into master and I have review bandwidth now so I think we can get this moving quickly.

I've left some minor review comments below. Once you've had a chance to address these (and maybe merge master one last time) I should be able to return and give a more detailed review quickly.

peabrainiac committed:
merge master

peabrainiac committed:
apply suggestions

peabrainiac commented on Mathlib/Geometry/Diffeology/Basic.lean:
```diff
@@ -0,0 +1,479 @@
+  /-- The plots `EuclideanSpace ℝ (Fin n) → X` representing the smooth ways to map
+  `EuclideanSpace ℝ (Fin n)` into `X`. This is the main
+  piece of data underlying the diffeology. -/
+  plots (n : ℕ) : Set (EuclideanSpace ℝ (Fin n) → X)
```
The main reason is that to relate the two equivalent definitions of diffeological spaces in terms of plots on $\mathbb R^n$ / open subsets of $\mathbb R^n$, one needs something like a neighbourhood basis of open sets that are diffeomorphic to $\mathbb R^n$ around every $x\in\mathbb R^n$ - I think in theory $p$-balls would work for this for any $p$, but as far as I know the only homeomorphism we have actually written down for this in mathlib is only smooth for $p=2$ (see [OpenPartialHomeomorph.contDiff_univUnitBall](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/InnerProductSpace/Calculus.html#OpenPartialHomeomorph.contDiff_univUnitBall)). You can see how I used it in the proof of `DiffeologicalSpace.ofCorePlotsOn` (line 325), and there are a few more proofs like it in files I have not yet PRed.

ocfnash replied:
I see.

Indeed on reflection there's nothing superior about `Fin n → ℝ` over `EuclideanSpace ℝ (Fin n)` in this setting: our notion of continuous differentiability demands a norm (which is the $p = ∞$ norm for `Fin n → ℝ`). I suppose the most natural thing to do would be not to make a choice and work with all finite-dimensional normed real vector spaces but quantifying over these will encounter universe issues so we need to pick a skeleton and I'd say it's a tie between $p = ∞$ and $p = 2$.

So although the fact that, at this moment in time we have only proved smoothness for [OpenPartialHomeomorph.contDiff_univUnitBall](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/InnerProductSpace/Calculus.html#OpenPartialHomeomorph.contDiff_univUnitBall) for $p = 2$, is not a good reason to found an entire theory in one way or another, I think it is OK to use it to break a tie.

peabrainiac commented on Mathlib/Geometry/Diffeology/Basic.lean:
```diff
@@ -0,0 +1,479 @@
+      ∀ {m : ℕ} {f : EuclideanSpace ℝ (Fin m) → EuclideanSpace ℝ (Fin n)},
+        (hfu : ∀ x, f x ∈ u) → ContDiff ℝ ∞ f → p ∘ f ∈ plots m) → p ∈ plots n
+  /-- The D-topology of the diffeology. This is included as part of the data in order to give
+  control over what the D-topology is defeq to. -/
```
I added the comment, but out of curiosity - can you point me to where those notes actually are? I've seen the notes page in the mathlib3 docs, but don't know where it got moved to in lean 4. Last I looked I couldn't find it and thought it was removed.

ocfnash commented on Mathlib/Geometry/Diffeology/Basic.lean:
```diff
@@ -0,0 +1,478 @@
+      Set.preimage_inter.symm ▸ (IsOpen.inter (hs p hp) (ht p hp))
+    isOpen_sUnion := fun _ hs _ p hp ↦
+      Set.preimage_sUnion ▸ isOpen_biUnion fun u hu ↦ hs u hu p hp
+  }
```
```suggestion
      Set.preimage_sUnion ▸ isOpen_biUnion fun u hu ↦ hs u hu p hp }
```

ocfnash reviewed (commented):
I've left a few more style points below. This is definitely converging to something nice but I'll probably have another few rounds of such nitpicks before we hit merge.

peabrainiac committed:
use macro `𝔼ⁿ` for `EuclideanSpace ℝ (Fin n)`

peabrainiac committed:
some smaller improvements

peabrainiac committed:
move to `Diffeology` namespace

peabrainiac commented on Mathlib/Geometry/Diffeology/Basic.lean:
```diff
@@ -0,0 +1,478 @@
+  /-- The D-topology consists of exactly those sets whose preimages under plots are all open. -/
+  isOpen_iff_preimages_plots {u : Set X} : dTopology.IsOpen u ↔
+      ∀ {n : ℕ}, ∀ p ∈ plots n, IsOpen (p ⁻¹' u) := by rfl
+
```
I don't particularly like `DiffeologicalSpace` as a namespace name because it makes it less clear whether `DiffeologicalSpace.xyz` is a lemma specifically about the `DiffeologicalSpace` structure and meant for dot notation or just a general lemma on diffeology. But I understand that namespacing this is probably a good idea, so I've moved most declarations to a new namespace `Diffeology` somewhat analogous to `Topology` in mathlib now.

ocfnash commented on Mathlib/Geometry/Diffeology/Basic.lean:
```diff
@@ -0,0 +1,470 @@
+Necessary because the D-topologies on for example products and subspaces don't agree with
+the product and subspace topologies. -/
+class IsDTopCompatible (X : Type*) [t : TopologicalSpace X] [DiffeologicalSpace X] : Prop where
+  dTop_eq : DTop = t
```
It seems to me that you should not need this class. The whole point of making `DiffeologicalSpace` include the data of `TopologicalSpace` is to give us the ability to solve this problem.

To be precise, when defining the `DiffeologicalSpace` structure for a subtype / product type / pi type, you would define it to have the corresponding topology. The only cost will be that in these cases `isOpen_iff_preimages_plots` stops being try by `rfl` and a real proof is required. This is exactly what we do for the corresponding situation for metric spaces, for example.

peabrainiac committed:
remove notations for explicitly handling diffeologies

peabrainiac committed:
rename `euclideanDiffeology` to `NormedSpace.toDiffeology`

peabrainiac commented on Mathlib/Geometry/Diffeology/Basic.lean:
```diff
@@ -0,0 +1,478 @@
+notation (name := IsPlot_of) "IsPlot[" d "]" => @IsPlot _ d
+
+/-- Notation for smoothness with respect to non-standard diffeologies. -/
+notation (name := DSmooth_of) "DSmooth[" d₁ ", " d₂ "]" => @DSmooth _ _ d₁ d₂
```
That seems very reasonable; I have removed them now.

peabrainiac committed:
rename `DTop` to `dTopology`

ocfnash committed:
Final style tweaks

With the author's permission I am pushing up a commit with
minor suggestions, rather than sticking to what would otherwise
be quite a painful experience in the GitHub UI, given the size of
the PR.  Many of these are subjective but I think worth having.
